File: 20. RNA.R
## Redes neuronales artificiales ##


# libraries ---------------------------------------------------------------

library(tidyverse)
library(caTools)
library(h2o)

# read data ---------------------------------------------------------------

datos <- readr::read_csv("data/Churn_Modelling.csv")

datos <- datos %>% 
  dplyr::select(-(1:3)) %>% 
  dplyr::mutate(Geography = factor(Geography, 
                                   levels = c("France", 
                                              "Spain", "Germany"),
                                   labels = c(1, 2, 3)),
                Gender = factor(Gender, 
                                levels = c("Female", "Male"),
                                labels = c(0, 1)),
                Exited = factor(Exited))

str(datos)

# escalado de variables ---------------------------------------------------
# No es necesario escalar ############
# datos <- datos %>% 
#   dplyr::mutate(across(-11, scale)) %>% 
#   purrr::map_df(function(x) {attributes(x) <- NULL; x}) %>% 
#   purrr::modify_at(11, factor)

# Entrenamiento y validación ----------------------------------------------

set.seed(123)

split <- caTools::sample.split(datos$Exited, SplitRatio = 0.75)

entrenamiento <- datos %>% 
  subset(split == T)

validacion <- datos %>% 
  subset(split == F)


# ajustar clasificador ----------------------------------------------------

h2o.init(nthreads = 3)

h2o_train <- as.h2o(as.data.frame(entrenamiento))
h20_test <- as.h2o(as.data.frame(validacion))

clasificador <- h2o.deeplearning(y = "Exited",
                                 training_frame = h2o_train,
                                 activation = "Rectifier",
                                 hidden = c(6, 6),
                                 epochs = 100,
                                 train_samples_per_iteration = -2,
                                 standardize = TRUE) # no es necesario escalar

# predicción de resultados ------------------------------------------------

pred <- h2o.predict(clasificador, h20_test) %>% 
  as.data.frame()

y_pred <- pred$predict

# matriz confusion --------------------------------------------------------

table(validacion$Exited, y_pred)

sum(diag(table(validacion$Exited, y_pred)))/sum(table(validacion$Exited, y_pred))

# cerrar la sesión de h2o -------------------------------------------------

h2o.shutdown(prompt = FALSE)

File: 20. RNA.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 22 10:45:24 2022

@author: Admin
"""

## Redes neuronales artificiales ##

# =============================================================================
# Libraries
# =============================================================================

import numpy as np
import pandas as pd
import sklearn.preprocessing as sklp
import sklearn.model_selection as sklm
import sklearn.compose as sklc
import sklearn.metrics as sklmetric

# =============================================================================
# Read data
# =============================================================================

churn = pd.read_csv("data\Churn_Modelling.csv")

x = churn.iloc[:, 3:13].values
y = churn.iloc[:, 13].values


# =============================================================================
# Datos categóricos
# =============================================================================

label_encoder_x1 = sklp.LabelEncoder()
x[:, 1] = label_encoder_x1.fit_transform(x[:, 1])

label_encoder_x2 = sklp.LabelEncoder()
x[:, 2] = label_encoder_x2.fit_transform(x[:, 2])

ct = sklc.ColumnTransformer(
    [("one_hot_encoder", sklp.OneHotEncoder(categories = "auto"), [1])],
    remainder = "passthrough")

x = np.array(ct.fit_transform(x), dtype = np.float64)

x = x[:,1:]

# =============================================================================
# Escalado de variables
# =============================================================================

sc_x = sklp.StandardScaler()
x = sc_x.fit_transform(x)

# =============================================================================
# División de entrenamiento y test
# =============================================================================

x_train, x_test, y_train, y_test = sklm.train_test_split(x, y, test_size= 0.25,
                                                         random_state = 0)

# =============================================================================
# Importar librerías adicionales
# =============================================================================

import keras
from keras.models import Sequential # inicializar RNA
from keras.layers import Dense

# =============================================================================
# Inicializar la red neuronal
# =============================================================================

clasificador = Sequential()

# Añadir capas de entrada y capas ocultas

clasificador.add(Dense(units = 6, kernel_initializer = "uniform",
                       activation = "relu", input_dim = 11))

## Añadir segunda capa oculta
clasificador.add(Dense(units = 6, kernel_initializer = "uniform",
                       activation = "relu"))

## Añadir capa de salida

clasificador.add(Dense(units = 1, kernel_initializer = "uniform",
                       activation = "sigmoid"))

# =============================================================================
# Ajustar RNA
# =============================================================================
  
clasificador.compile(optimizer = "adam", loss = "binary_crossentropy",
                     metrics = "accuracy")


clasificador.fit(x_train, y_train, batch_size = 10, epochs = 100)

# =============================================================================
# Predicción de resultados 
# =============================================================================

y_pred = clasificador.predict(x_test)

y_pred = (y_pred>0.5)

# =============================================================================
# Matriz de confusion
# =============================================================================

cm = sklmetric.confusion_matrix(y_test, y_pred)
cm
